CTN-1033 custom changes for creatives

tugis opened the pull request:
- Implementing logic to invalidate cache keys
- Ajusting some tests
- Adjusting logic to follow RESTful API principle

tugis committed:
CTN-1033 custom changes for creatives
- Implementing logic to invalidate cache keys
- Ajusting some tests
- Adjusting logic to follow RESTful API principle

mloureiro commented on src/CacheProfiles/CacheAllSuccessfulGetRequests.php:
```diff
@@ -38,4 +38,14 @@ public function shouldCacheResponse(Response $response)
+     */
+    public function shouldInvalidateCache(Request $request)
+    {
+        return $request->isMethod('patch') || $request->isMethod('delete') || $request->isMethod('post');
```
Why not `return !$request->isMethod('get');` ?

tugis replied:
Hmm... the reason I'm doing this is because this way, I have control over the verbs I want to handle. There are other HTTP verbs which we don't handle.

mloureiro commented on src/RequestHasher.php:
```diff
@@ -27,7 +27,7 @@ public function __construct(CacheProfile $cacheProfile)
     {
         return 'laravel-responsecache-'.md5(
-            "{$request->getUri()}/{$request->getMethod()}/".$this->cacheProfile->cacheNameSuffix($request)
+            "{$request->getUri()}/".$this->cacheProfile->cacheNameSuffix($request)
```
please add surround concatenation (`.`) with spaces please
and as personal opinion I really prefer 
```php
sprintf('%s/%s', $request->getUri(), $this->cacheProfile->cacheNameSuffix($request))
```

mloureiro commented on src/ResponseCache.php:
```diff
@@ -61,6 +61,28 @@ public function shouldCache(Request $request, Response $response)
+            return false;
+        }
+
+        return true;
```
😅 
What about:
```php
return config('laravel-responsecache.enabled') 
    && ! $request->attributes->has('laravel-cacheresponse.doNotCache')
    && $this->cacheProfile->shouldCacheRequest($request);
```
**Notice** that it will not run the next check/validation if the previous fails.

--- 
also is this correct `'laravel-cacheresponse.doNotCache'` ?
- `cacheresponse` as no divisor between the 2 words neither camelCase
- `laravel-cacheresponse` lowercase with dashes to separate words
- `doNotCache` camelCase

tugis replied:
>also is this correct 'laravel-cacheresponse.doNotCache' ?

This is used as you can see in the original module, so yes, I suppose it works.

mloureiro replied:
> This is used as you can see in the original module, so yes, I suppose it works.

gotcha, but is weird :rofl:

tugis committed:
CTN-1033 changes according to PR review

tugis committed:
CTN-1033 removing middleware cache invalidations
We decided after discussion to do cache invalidations at Repository level instead of middleware level (where it's difficult to find out if the cache should be invalidated or not).

tugis committed:
CTN-1033 changing cache key construction
With this change we can identify the service the cache key belongs to. Also, removing full uri from it (useful for cache invalidation).

tugis committed:
CTN-1033 passing cache suffix
Implementing possibility of cache suffix override so we can pass `userId` as suffix for cache invalidation

tugis committed:
CTN-1033 removing caching of redirect requests
Redirect requests (304) are client specific and so cannot be cached, otherwise other clients will get an empty response

tugis committed:
CTN-1033 sending 304 when response is not modified
